src/main/scala/com/fulcrumgenomics/util/ReadStructure.scala

```diff
@@ -0,0 +1,143 @@
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.util
+
+import com.fulcrumgenomics.util.BaseSymbol._
+import com.fulcrumgenomics.util.ReadSegment._
+
+object ReadStructure {
+  private val segmentPattern = s"\\d+[${SymbolChars.mkString}]".r
+  private val readStructurePattern = s"(\\d+[${SymbolChars.mkString}])+".r
+
+  /** Creates a read structure from a string */
+  def apply(readStructure: String): ReadStructure = {
+    if (readStructurePattern.findAllIn(readStructure).mkString.length != readStructure.length) {
+      throw new IllegalArgumentException("Read structure was not valid: " + readStructure)
+    }
+    val segments = segmentPattern.findAllIn(readStructure).toSeq
+    val lengths = segments.map(seg => Integer.parseInt(seg.substring(0, seg.length-1)))
+    new ReadStructure(segs = segments.zipWithIndex.map {
+      case (seg, idx) =>
+        val offset = lengths.slice(0, idx).sum
+        val length = lengths(idx)
+        if (length == 0) throw new IllegalArgumentException("Read structure had a segment with zero length: " + seg)
+        seg.substring(seg.length-1).head match {
+          case TemplateSymbol.symbol         => new Template(offset=offset, length=length)
+          case SampleBarcodeSymbol.symbol    => new SampleBarcode(offset=offset, length=length)
+          case MolecularBarcodeSymbol.symbol => new MolecularBarcode(offset=offset, length=length)
+          case SkipSymbol.symbol             => new Skip(offset=offset, length=length)
+          case s                             => throw new IllegalArgumentException("Read segment type unknown: " + s)
+        }
+    }.toTraversable)
+  }
+
+  /** Splits the given bases into tuples with its associated read segment. */
+  def extract(bases: String, readStructure: ReadStructure, symbol: Option[BaseSymbol] = None, strict: Boolean = true): Seq[(String, ReadSegment)] = {
```

Comment:
1. Unstrict would allow you to pass in 50bp read and a read structure of `75T50T`.  The second segment is filtered out in this case.  Strict would throw an exception.
2. Unstrict would allow you to pass in 50bp read and a read structure of `75T`.  Strict would throw an exception.
I think you object to (1) but not (2).  

If you think about each read from a pair independently I agree with you, since I think in all cases the last segment would be the template bases.  For index reads I agree too.  But for a read structure that describes the pair jointly (see `IlluminaBasecallsToSam`) then this may not be true.  Consider if the second read of the pair fails completely or is completely trimmed.  We may never use `ReadStructure` across multiple reads (ex. for a whole template) in `fgbio`, and if so, I agree with you and we should document that clearly.  What's your thought?